servers/rendering/renderer_scene_cull.cpp

```diff
@@ -3500,6 +3589,248 @@ TypedArray<Image> RendererSceneCull::bake_render_uv2(RID p_base, const Vector<RI
+#undef CHECK_OFFSET
+			}
+		}
+	}
+}
+
+bool RendererSceneCull::RendererSceneOcclusionCull::HZBuffer::is_occluded(const float p_bounds[6], const Vector3 &p_cam_position, const Transform &p_cam_inv_transform, const CameraMatrix &p_cam_projection, float p_near) const {
+	if (is_empty()) {
+		return false;
+	}
+
+	Vector3 closest_point = Vector3(CLAMP(p_cam_position.x, p_bounds[0], p_bounds[3]), CLAMP(p_cam_position.y, p_bounds[1], p_bounds[4]), CLAMP(p_cam_position.z, p_bounds[2], p_bounds[5]));
+
+	if (closest_point == p_cam_position) {
+		return false;
+	}
+
+	Vector3 closest_point_view = p_cam_inv_transform.xform(closest_point);
+	if (closest_point_view.z > -p_near) {
+		return false;
+	}
+
+	float min_depth;
+	if (p_cam_projection.is_orthogonal()) {
+		min_depth = (-closest_point_view.z) - p_near;
+	} else {
+		float r = -p_near / closest_point_view.z;
+		Vector3 closest_point_proj = Vector3(closest_point_view.x * r, closest_point_view.y * r, -p_near);
+		min_depth = closest_point_proj.distance_to(closest_point_view);
+	}
+
+	Vector2 rect_min = Vector2(FLT_MAX, FLT_MAX);
+	Vector2 rect_max = Vector2(FLT_MIN, FLT_MIN);
+
+	for (int j = 0; j < 8; j++) {
+		Vector3 c = RendererSceneCull::RendererSceneOcclusionCull::HZBuffer::corners[j];
+		Vector3 nc = Vector3(1, 1, 1) - c;
+		Vector3 corner = Vector3(p_bounds[0] * c.x + p_bounds[3] * nc.x, p_bounds[1] * c.y + p_bounds[4] * nc.y, p_bounds[2] * c.z + p_bounds[5] * nc.z);
+		Vector3 view = p_cam_inv_transform.xform(corner);
+
+		Vector3 projected = p_cam_projection.xform(view);
+		Vector2 normalized = Vector2(projected.x * 0.5f + 0.5f, projected.y * 0.5f + 0.5f);
+		rect_min = rect_min.min(normalized);
+		rect_max = rect_max.max(normalized);
+	}
+
+	rect_max = rect_max.min(Vector2(1, 1));
+	rect_min = rect_min.max(Vector2(0, 0));
+
+	int mip_count = mips.size();
+
+	Vector2 screen_diagonal = (rect_max - rect_min) * sizes[0];
+	float size = MAX(screen_diagonal.x, screen_diagonal.y);
+	float l = Math::ceil(Math::log2(size));
+	int lod = CLAMP(l, 0, mip_count - 1);
+
+	const int max_samples = 512;
+	int sample_count = 0;
+	bool visible = true;
+
+	for (; lod >= 0; lod--) {
+		int w = sizes[lod].x;
+		int h = sizes[lod].y;
+
+		int minx = CLAMP(rect_min.x * w - 1, 0, w - 1);
+		int maxx = CLAMP(rect_max.x * w + 1, 0, w - 1);
+
+		int miny = CLAMP(rect_min.y * h - 1, 0, h - 1);
+		int maxy = CLAMP(rect_max.y * h + 1, 0, h - 1);
+
+		sample_count += (maxx - minx + 1) * (maxy - miny + 1);
+
+		if (sample_count > max_samples) {
+			return false;
+		}
+
+		visible = false;
+		for (int y = miny; y <= maxy; y++) {
+			for (int x = minx; x <= maxx; x++) {
+				float depth = mips[lod][y * w + x];
+				if (depth > min_depth) {
+					visible = true;
+					break;
+				}
+			}
+			if (visible) {
+				break;
+			}
+		}
+
+		if (!visible) {
+			return true;
+		}
+	}
+
+	return !visible;
+}
+
+RID RendererSceneCull::RendererSceneOcclusionCull::HZBuffer::get_debug_texture() {
```

Comment:
I would probably avoid creating this every frame, if this is what you are doing, you can keep it cached.